### DISCUSSION TITLE
Waiting for new result in middleware

### DISCUSSION THREAD

**Discussion by grizzm0** [algolia/instantsearch#5035] [Q&A]
I want to read the new hits after the state changes. But when changing sort order onStateChange is triggered before the new hits is available. 

The code below gives two different results as the first console.log is fired before the hits has been updated and the second one after.

Is there any way to wait for the new hits without just guessing how long the response will take?

```javascript
function foobarMiddleware({ instantSearchInstance }) {
  return {
    onStateChange({uiState}) {
      console.log(instantSearchInstance.mainIndex.getResults().hits);

      setTimeout(() => {
        console.log(instantSearchInstance.mainIndex.getResults().hits);
      }, 500);
    },
  };
}

```

**Comment by Haroenv** [comment]
What is the goal of the middleware? It's not designed to react on the result at the moment (that would be more a widget), so I'd love to read what you're trying to achieve first

**Reply by grizzm0** [reply]
I need to add all our products to the GTM data layer. The insights middleware kind of does this. But just 20 at a time. I need to add all of them.

But perhaps I should do that in my custom widget that renders the results instead?

**Reply by Haroenv** [reply]
The view event for hits indeed is chunked in batches of 20, as there were several services that had the limit of 20 events per request to make the integration simpler.

Is there a technical reason why the products could not be sent in batches to GTM either? Maybe that's a simpler option than changing how events are sent.

If you already have a custom hits widget, you can of course send the even directly there without the indirection
